javascript/ql/src/Declarations/UnusedVariable.ql

```diff
@@ -94,36 +94,99 @@ predicate isEnumMember(VarDecl decl) {
-      not mentionedInJSDocComment(v) and
-      // exclude variables used to filter out unwanted properties
-      not isPropertyFilter(v) and
-      // exclude imports of React that are implicitly referenced by JSX
-      not isReactImportForJSX(v) and
-      // exclude names that are used as types
-      not isUsedAsType(vd) and
-      // exclude names that are used as namespaces from inside a type
-      not isUsedAsNamespace(vd) and
-      // exclude decorated functions and classes
-      not isDecorated(vd) and
-      // exclude names of enum members; they also define property names
-      not isEnumMember(vd) and
-      // ignore ambient declarations - too noisy
-      not vd.isAmbient()
-select vd, "Unused " + describe(vd) + "."
+/**
+ * An import statement that provides variable declarations.
+ */
+class ImportVarDeclProvider extends Stmt {
+
+  ImportVarDeclProvider() {
+    this instanceof ImportDeclaration or
+    this instanceof ImportEqualsDeclaration
+  }
+
+  /**
+   * Gets a variable declaration of this import.
+   */
+  VarDecl getAVarDecl() {
+    result = this.(ImportDeclaration).getASpecifier().getLocal() or
+    result = this.(ImportEqualsDeclaration).getId()
+  }
+}
+
+/**
+ * Holds if it is acceptable that `v` is unused.
+ */
+predicate whitelisted(UnusedLocal v) {
+  // exclude variables mentioned in JSDoc comments in externs
+  mentionedInJSDocComment(v) or
+  // exclude variables used to filter out unwanted properties
+  isPropertyFilter(v) or
+  // exclude imports of React that are implicitly referenced by JSX
+  isReactImportForJSX(v) or
+  // exclude names that are used as types
+  exists (VarDecl vd |
+    v = vd.getVariable() |
+    isUsedAsType(vd) or
+    // exclude names that are used as namespaces from inside a type
+    isUsedAsNamespace(vd)or
+    // exclude decorated functions and classes
+    isDecorated(vd) or
+    // exclude names of enum members; they also define property names
+    isEnumMember(vd) or
+    // ignore ambient declarations - too noisy
+    vd.isAmbient()
+  )
+}
+
+/**
+ * Holds if `vd` declares an unused variable that does not come from an import statement, as explained by `msg`.
+ */
+predicate unusedNonImports(VarDecl vd, string msg) {
+  exists (UnusedLocal v |
+    v = vd.getVariable() and
+    msg = "Unused " + describeVarDecl(vd) + "." and
+    not vd = any(ImportVarDeclProvider p).getAVarDecl() and
+    not whitelisted(v)
+  )
+}
+
+/**
+ * Holds if `provider` declares one or more unused variables, as explained by `msg`.
+ */
+predicate unusedImports(ImportVarDeclProvider provider, string msg) {
+  exists (string plural, string names |
+    names = concat(UnusedLocal v |
+      v = provider.getAVarDecl().getVariable() and
+      not whitelisted(v) |
+      v.getName(),
+      ", " order by v.getName()
+    ) and
+    not names = "" and
+    (
+      if count(UnusedLocal v |
+        v = provider.getAVarDecl().getVariable() and
+        not whitelisted(v)) > 1 then
```

Comment:
Could you perhaps pull this out into a predicate and reuse above?

Reply:
Perhaps make it a member predicate of `ImportVarDeclProvider`.

Reply:
Done. Should I squash the fixup?